Add option to split MC-table events into separate events by time

sm2909 opened the pull request:
for issue #523 

Clustering was supposed to be based on position and time simultaneously.
To achieve this, I implemented a Spatio-Temporal DBSCAN (ST-DBSCAN) algorithm, which includes a point in a cluster only when both spatial and temporal constraints are satisfied.

Traditional DBSCAN uses a Euclidean metric, which mixes position and time dimensions. This could incorrectly group points that are spatially close but temporally far apart (for example, position inside but time outside the threshold, and still euclidean distance inside threshold). The new approach avoids that issue by handling both constraints explicitly.

That is why I have used this new algorithm. Any prior implementation of this did not exist in julia libraries, so I have coded it from scratch

A new dependency Neighbors.jl was added in project.toml for this algorithm

`
tc = SolidStateDetectors.cluster_detector_hits(t, 20u"μm", 100u"s")
`

When this is run, clustering will happen with time constraint as well. Cluster properties (pos, edep) are calculated as before with the addition of thit as well.

```
tc.thit[1]
3-element view(::Vector{Quantity{Float32, 𝐓, Unitful.FreeUnits{(s,), 𝐓, nothing}}}, 1:3) with eltype Quantity{Float32, 𝐓, Unitful.FreeUnits{(s,), 𝐓, nothing}}:
 1.7250684f6 s
 1.7992052f6 s
 1.7991948f6 s
```

(This was run in the same case with which you demonstrated me in previous comment)


The problem that it has is that it is taking a lot of computer resources to compute the clusters. I tried using the best approaches (KDTree, inrange from Neighbors.jl), but the issue is arising due to the complexity of the problem as we have to simultaneously cluster based on position and time.

This is the computer resources typically being used when clustering using cluster_detector_hits function:

`133.839229 seconds (35.63 M allocations: 1.751 GiB, 0.57% gc time, 99.71% compilation time)`

(There might be a better approach or performance gains to be found. I tried my best and came up with this.)

Open for any modifications, which might be required for it to be integrated with whole project.

fhagemann commented:
Thanks for the initiative, I will look into this in more detail once I find a quite moment.
Maybe a quick question to get an order of magnitude: the `133.839229 seconds` you quote is for how many events/rows of the initial table?

sm2909 commented:
1000 rows

These are the exact commands that I ran:

```
using SolidStateDetectors, Geant4, Unitful
sim = Simulation{Float32}(SSD_examples[:InvertedCoax])
source = IsotopeSource(90, 228, 0.0, 0.0, CartesianPoint(0.05,0,0.05))
app = G4JLApplication(sim, source)
t = run_geant4_simulation(app, 1_000)
@time tc = SolidStateDetectors.cluster_detector_hits(t, 20u"μm", 100u"s")
```

fhagemann commented:
I see, thanks!
So, what I'm seeing is that these changes do two things:
- First, updating the current method of `cluster_detector_hits` to the new implementation of `CartesianPoint` (#489), which on the current `main` branch does not seem to be working with the Geant4 output.
- Update the clustering algorithm to cluster both time and/or positions

Would it be very cumbersome for you to split these two things into two commits, to see what exactly the changes to the clustering algorithm are?

And whether the long run-time of 2 minutes results from the updates to the new `CartesianPoints` (maybe benchmark this before implementing the timing clustering and see how long it takes to run for 1000 events) or the new clustering?

fhagemann commented on test/test_geant4.jl:
```diff
@@ -66,6 +67,11 @@ end
     @test length(evts) == 100
 
+    # Cluster events by radius and time
+    @test_nowarn clusted_evts = SolidStateDetectors.cluster_detector_hits(evts, 10u"µm", 100u"s")
```
I'm sorry, my previous message had a typo which is causing these tests to fail :sweat_smile: 
```suggestion
    @test_nowarn clustered_evts = SolidStateDetectors.cluster_detector_hits(evts, 10u"µm", 100u"s")
```

fhagemann commented on src/ChargeClustering/ChargeClustering.jl:
```diff
@@ -44,28 +51,36 @@ function cluster_detector_hits(
                 else
                     weights = ustrip.(c_hits.edep) .* inv(esum)
                     push!(r_pos, sum(c_hits.pos .* weights))
+                    push!(r_thit, sum(c_hits.thit .* weights))
```
You could ensure to convert `r_pos` back to its original type (`StaticVector` or `CartesianPoint`) when returning this.

fhagemann replied:
@oschulz ?

fhagemann commented on src/ChargeClustering/ChargeClustering.jl:
```diff
@@ -8,31 +8,38 @@ function cluster_detector_hits(
         @assert all(isequal(d_detno), d_hits.detno)
         if length(d_hits) > 3
-            clusters = Clustering.dbscan(ustrip.(flatview(d_hits.pos)), ustripped_cradius, leafsize = 20, min_neighbors = 1, min_cluster_size = 1)
+            clusters = st_dbscan(hcat(ustrip.(d_hits.pos)...)', ustrip.(d_hits.thit), eps_space = ustripped_cradius, eps_time = ustripped_ctime, min_cluster_size = 1)
```
Maybe instead of converting everything first to an `SVector` and then to this `Matrix` for `st_dbscan`, we might consider defining a function that defines this matrix directly from `grouped` or `d_hits.pos`

Could be something like (I did not test this, just wrote this out in this comment as a starting point):
```julia
function create_dbscan_matrix(pos::AbstractVector{CartesianPoint{T}})
    ustrip.(hcat(getfield.(pos, :x), getfield.(pos, :y), getfield.(pos, :z)))'
end

function create_dbscan_matrix(pos::AbstractVector{StaticVector{T}})
    ustrip.(hcat(d_hits.pos)...))'
end

fhagemann commented on src/ChargeClustering/ChargeClustering.jl:
```diff
@@ -130,3 +145,78 @@ function group_points_by_distance(pts::AbstractVector{CartesianPoint{T}}, energi
+        pos = hcat(pos...)'   # convert Vector of SVector{3,T} to Nx3 Matrix
+    elseif !(pos isa AbstractMatrix{<:AbstractFloat})
+        error("Unsupported position format: must be a matrix of floats or vector of points")
+    end
```
Do we still need this or is this handled is `cluster_detector_hits` already.
Because you require your input to this method to already be an `AbstractMatrix`.

fhagemann commented on src/ChargeClustering/ChargeClustering.jl:
```diff
@@ -130,3 +145,78 @@ function group_points_by_distance(pts::AbstractVector{CartesianPoint{T}}, energi
+    end
+
+    # KDTree for spatial queries
+    tree = NearestNeighbors.KDTree(pos'; leafsize = leafsize)
```
Your transposing `pos` in `cluster_detector_hits` when passing this to `st_dbscan`, but transpose this here again. Can this be avoided?

fhagemann reviewed (commented):
Here are some comments.

I'm wondering here: would it maybe be more efficient to 
1. split all events by time (separate a row into multiple rows if the hits are more than `eps_time` apart). I expect most of the `Geant4` hits to have identical/very similar `thit` entries if they resulted from the same nuclear decay.
2. then, perform the spatial clustering like we used to before (or the more efficient implementation)?

Based on your experience in implementing this, do you think that this would be faster?
Or do you have some feeling for what is limiting the runtime of this clustering?

sm2909 commented:
> Here are some comments.
> 
> I'm wondering here: would it maybe be more efficient to
> 
> 1. split all events by time (separate a row into multiple rows if the hits are more than `eps_time` apart). I expect most of the `Geant4` hits to have identical/very similar `thit` entries if they resulted from the same nuclear decay.
> 2. then, perform the spatial clustering like we used to before (or the more efficient implementation)?
> 
> Based on your experience in implementing this, do you think that this would be faster? Or do you have some feeling for what is limiting the runtime of this clustering?

That would lead to incorrect clustering. Consider this toy example:

eps_time = 10s
eps_pos = 10m
(consider position 1d here for easier visualization)

hit 1: 1s, 1m
hit 2: 10s, 100m
hit 3: 15s, 5m

First clustering by time, all hits are in same cluster (eps_time < 10).
Then cluster based on position, [1, 3] are a cluster and [2] is another cluster.
But [1, 3] should not belong to same cluster as eps_time > 10 between them.

(Earlier hit 2 formed the common link between hit 1 and 3 for eps_time. But as finally 2 is not in the same cluster, hit 1 and 3 should not be kept in the same cluster.)

fhagemann commented:
> That would lead to incorrect clustering. Consider this toy example:
> 
> eps_time = 10s eps_pos = 10m (consider position 1d here for easier visualization)
> 
> hit 1: 1s, 1m hit 2: 10s, 100m hit 3: 15s, 5m
> 
> First clustering by time, all hits are in same cluster (eps_time < 10). Then cluster based on position, [1, 3] are a cluster and [2] is another cluster. But [1, 3] should not belong to same cluster as eps_time > 10 between them.
> 
> (Earlier hit 2 formed the common link between hit 1 and 3 for eps_time. But as finally 2 is not in the same cluster, hit 1 and 3 should not be kept in the same cluster.)

I see what you mean.
The idea behind time clustering would be to separate events that are more than a given `eps_time` apart. This is because, in our pulse shape simulation, we start drifting all charge clouds at the same initial time.
Other than with distance, I would also like to make sure that grouped events are never more than `eps_time` apart.

So, even if the hits were 
```
hit 1: 1s, 1m 
hit 2: 10s, 1m
hit 3: 15s, 1m
```
and `eps_time` was `10s`, I would like to NEVER group hit 1 and hit3.

Let`s assume this toy model:
```
for N in 1:1000
hit (N): (N) ns, 1m

```
If I choose `eps_time = 1u"ns"` because my simulation step is `1u"ns"`, I would group all hits within the `1u"µs"` together, even though this might not be what I want. I, at most, would like the first two hits to be combined and start a new group with the third hit etc. 

Just to think out loud (doesn't mean that this is the best way to go): the time grouping could either start with the first hit and combine everything that's withing `eps_time` after the first hit.
OR we can sort all events by time and start grouping with the two hits that are closest together, group them, and then keep on adding more events if they are withing `eps_time` or start with new events afterwards.

The time clustering should be there to make sure that we only drift hits together, if they actually coincide "enough" in time, whereas the space clustering makes sure that we do not simulate thousands of small charge clouds (which might be the output of Geant4), especially if they are all very close together.

sm2909 commented:
> > That would lead to incorrect clustering. Consider this toy example:
> > eps_time = 10s eps_pos = 10m (consider position 1d here for easier visualization)
> > hit 1: 1s, 1m hit 2: 10s, 100m hit 3: 15s, 5m
> > First clustering by time, all hits are in same cluster (eps_time < 10). Then cluster based on position, [1, 3] are a cluster and [2] is another cluster. But [1, 3] should not belong to same cluster as eps_time > 10 between them.
> > (Earlier hit 2 formed the common link between hit 1 and 3 for eps_time. But as finally 2 is not in the same cluster, hit 1 and 3 should not be kept in the same cluster.)
> 
> I see what you mean. The idea behind time clustering would be to separate events that are more than a given `eps_time` apart. This is because, in our pulse shape simulation, we start drifting all charge clouds at the same initial time. Other than with distance, I would also like to make sure that grouped events are never more than `eps_time` apart.
> 
> So, even if the hits were
> 
> ```
> hit 1: 1s, 1m 
> hit 2: 10s, 1m
> hit 3: 15s, 1m
> ```
> 
> and `eps_time` was `10s`, I would like to NEVER group hit 1 and hit3.
> 
> Let`s assume this toy model:
> 
> ```
> for N in 1:1000
> hit (N): (N) ns, 1m
> ```
> 
> If I choose `eps_time = 1u"ns"` because my simulation step is `1u"ns"`, I would group all hits within the `1u"µs"` together, even though this might not be what I want. I, at most, would like the first two hits to be combined and start a new group with the third hit etc.
> 
> Just to think out loud (doesn't mean that this is the best way to go): the time grouping could either start with the first hit and combine everything that's withing `eps_time` after the first hit. OR we can sort all events by time and start grouping with the two hits that are closest together, group them, and then keep on adding more events if they are withing `eps_time` or start with new events afterwards.
> 
> The time clustering should be there to make sure that we only drift hits together, if they actually coincide "enough" in time, whereas the space clustering makes sure that we do not simulate thousands of small charge clouds (which might be the output of Geant4), especially if they are all very close together.


Thanks for clarifying the use-case of this algorithm. 

As we just want to bring together the hits "close enough" in space and time, we might not need dbscan algorithm at all. Rather it could be achieved with a lot simpler algorithms.

For the 1:1000 hits example that you gave, would give same cluster if we do it by dbscan, however we implement it. As long as the hits are dense in time (or space), it will keep on adding. It would lead to huge clusters which is not what we need.

Using the following algorithm would be better in solving our purpose (logic):

```
for each hit:
    if hit already in a cluster: continue
    search all other hits which are not already part of a cluster & lie with eps of hit
    if satisfies above condition: cluster with hit
end
```
 
The implementation for this would be straightforward as well which means runtime will be low.

fhagemann commented on src/ChargeClustering/ChargeClustering.jl:
And it also looks like we don't support `Clustering@0.14` since quite a while ago (#372),
so we can remove the method in line 5 that I introduced to support both `Clustering@0.14` (line 5) and `Clustering@0.15` (line 4) in #332

sm2909 commented on src/ChargeClustering/ChargeClustering.jl:
```diff
@@ -7,61 +7,106 @@
+        for tg in time_groups
+            t_hits = view(d_hits, tg)
+            if length(t_hits) > 3
+                clusters = Clustering.dbscan(ustrip.(flatview(t_hits.pos)), ustripped_cradius;
```
We could do this, but it would require a lot of work in the CartesianPoint type.

We need to implement:
getindex, length, iterate, arithmetic (+, /, *), dot, norm, Statistics.mean, matrix conversion
(all the changes are listed by chatGPT)

this is the bare minimum required to make the function work with CartesianPoint type without conversion.

Then also, it might not be most efficient, as SVector is best optimized at compiler level for the various functions that we are doing here.

sm2909 commented:
1. Added some more test: checking eltype of pos, sum of all edep, no NaN in output table
2. push! in previous has been moved to helper function to save allocations
3. pos type is converted to CartesianPoint in the function (with table input) to avoid errors.
4. Removed Clustering@0.14 supporter method as you suggested.

fhagemann commented:
Ok, sorry for making this chaotic, but in PR #558, I added code that would make sure that the OLD implementation of `cluster_detector_hits` would work and merged this into main.
This also included some clever way to deal with both `CartesianPoint` and `StaticVector` as `pos` column of the `Table`.
https://github.com/JuliaPhysics/SolidStateDetectors.jl/commit/d6a62f381b3e134060519d44b05a54be2a50fe72

Could you build your changes on top of this, and see if the tests require changes?

fhagemann commented:
And I think we might still need to tackle the issue with the runtime to avoid this from taking forever to run.
Check how fast it was before this change, after this change, and maybe with some default (`cluster_time = Inf * u"s"`) that should result in the same behavior as before, and how long that takes.

But we're getting there!

sm2909 commented:
Benchmarks before change:

```
 @time tc = SolidStateDetectors.cluster_detector_hits(t, 20u"μm")
  4.338386 seconds (6.55 M allocations: 384.269 MiB, 3.01% gc time, 96.09% compilation time)
```

Benchmarks after change (clustering by time as well):

```
@time tc = SolidStateDetectors.cluster_detector_hits(t, 20u"μm", Inf *u"s")
  5.322641 seconds (8.62 M allocations: 483.518 MiB, 6.93% gc time, 97.33% compilation time)
```

fhagemann commented on src/ChargeClustering/ChargeClustering.jl:
```diff
@@ -1,64 +1,117 @@
+    cluster_time::RealQuantity
+) where {TT<:RealQuantity, PT <: RealQuantity, TTT <: RealQuantity}
+
+    pos = [SVector(p.x, p.y, p.z) for p in pos]   # converting CartesianPoint to SVector
```
--> It is not needed anymore and I removed it in 830eca8

fhagemann reviewed (dismissed):
I pushed some minor changes (a typo bugfix, and support for both `CartesianPoint` and `SVector` as type of the `pos` column of the table).

It might make sense to do a final check that the time splitting performs as intended (also looking at energy spectra etc.) and adding this to the tests if wanted.]
And we might also (in a separate PR) look into performance in a bit more detail to further decrease the runtime of this if possible.

But in general, this PR looks good to me now!

@sm2909 thanks for the initiative!

fhagemann commented on src/ChargeClustering/ChargeClustering.jl:
```diff
@@ -1,64 +1,115 @@
 
-    (detno = r_detno, edep = r_edep, pos = r_pos)
+    (detno = r_detno, edep = r_edep, pos = r_pos, thit = r_thit)
 end
```
We could remove `time_groups` and `finalize_group!` by doing what the second `for`-loop does whenever you would `push!` something to `time_groups`:
```suggestion
function cluster_detector_hits(
    detno::AbstractVector{<:Integer},
    edep::AbstractVector{TT},
    pos::AbstractVector{<:Union{<:StaticVector{3,PT}, <:CartesianPoint{PT}}},
    thit::AbstractVector{TTT},
    cluster_radius::RealQuantity,
    cluster_time::RealQuantity
) where {TT<:RealQuantity, PT <: RealQuantity, TTT <: RealQuantity}

    unsorted = TypedTables.Table(detno = detno, edep = edep, pos = pos, thit = thit)
    sorting_idxs = sortperm(unsorted.detno)
    sorted = view(unsorted, sorting_idxs)
    grouped = TypedTables.Table(consgroupedview(sorted.detno, TypedTables.columns(sorted)))

    r_detno = similar(detno, 0)
    r_edep = similar(edep, 0)
    r_pos = similar(pos, 0)
    r_thit = similar(thit, 0)

    ustripped_cradius = ustrip(internal_length_unit, cluster_radius)

    for d_hits_nt in grouped

        d_hits = TypedTables.Table(d_hits_nt)

        # sort hits by time
        t_sort_idx = sortperm(d_hits.thit)
        d_hits = view(d_hits, t_sort_idx)
        t_vals = d_hits.thit

        current_group = 1
        @inbounds for i in eachindex(t_vals) .+ 1
            if i > lastindex(t_vals) || t_vals[i] - t_vals[current_group] > cluster_time
                t_hits = view(d_hits, current_group:i-1)
                if length(t_hits) > 3
                    d_detno = first(t_hits.detno)
                    @assert all(isequal(d_detno), t_hits.detno)
                    clusters = Clustering.dbscan(hcat((ustrip.(internal_length_unit, getindex.(t_hits.pos,i)) for i in 1:3)...)', 
                        ustripped_cradius, leafsize = 20, min_neighbors = 1, min_cluster_size = 1).clusters
                    
                    for c in clusters
                        idxs = vcat(c.boundary_indices, c.core_indices)
                        @assert length(idxs) == c.size
                        c_hits = view(t_hits, idxs)
                        esum = sum(c_hits.edep)

                        push!(r_detno, d_detno)
                        push!(r_edep, esum)
                        if esum ≈ zero(TT)
                            push!(r_pos, barycenter(c_hits.pos))
                            push!(r_thit, mean(c_hits.thit))
                        else
                            weights = ustrip.(Unitful.NoUnits, c_hits.edep .* inv(esum))
                            push!(r_pos, barycenter(c_hits.pos, StatsBase.Weights(weights)))
                            push!(r_thit, sum(c_hits.thit .* weights))
                        end
                    end
                else
                    append!(r_detno, t_hits.detno)
                    append!(r_edep, t_hits.edep)
                    append!(r_pos, t_hits.pos)
                    append!(r_thit, t_hits.thit)
                end
                current_group = i
            end
        end
    end

    (detno = r_detno, edep = r_edep, pos = r_pos, thit = r_thit)
end
```

sm2909 replied:
I have implemented these changes that you have suggested

`clusters = Clustering.dbscan(hcat((ustrip.(internal_length_unit, getindex.(t_hits.pos,i)) for i in 1:3)...)'`

in this line `internal_length_unit` was causing unit mismatch with cluster_radius, so I have used the `posunit` from cluster_radius while stripping units of `pos` and `cluster_radius` to maintain consistency

fhagemann commented on src/ChargeClustering/ChargeClustering.jl:
Nice! Can you also give numbers on how long this takes to run (maybe compared with the previous implementation?)

sm2909 replied:
works same for 0u"ns"

negative time cases are not handled
output table is generated in same runtime, but it is wrong

sm2909 replied:
For the conflicts:

For ChargeClustering.jl, all the changes from my branch should be taken as the changes in the main codebase does not use clustering based on time.

You will need to look in the tests file. There is some conflict between the changes you made on this branch versus the recent new code in the main codebase

fhagemann replied:
Do you want me to resolve all conflicts? :)

fhagemann replied:
I re-resolved the merge conflicts to account for this, let's see if that works now.

sm2909 commented on test/test_geant4.jl:
```diff
@@ -110,15 +142,20 @@ end
 
     # Check clustering with cluster radius being unitless or with the wrong unit
-    clustered_evts_static_unitless = SolidStateDetectors.cluster_detector_hits(evts_static, 1e-5)
+    clustered_evts_static_unitless = SolidStateDetectors.cluster_detector_hits(evts_static, 1e-5, 1u"ns")
```
In earlier version of tests, it still used the function where clustering was happening only based on space and not time. The function signature has changed now, so I have modified these tests here.
Also I added corresponding tests for incorrect units of cluster_time.

fhagemann commented:
Yes, this is why it's prefered to use `to_internal_units` on `CartesianPoint` (or `SVector`, depending on which it is).
I undid your merge commit and re-merged while resolving the merge conflicts.

fhagemann commented on src/ChargeClustering/ChargeClustering.jl:
```diff
@@ -77,19 +76,12 @@ function cluster_detector_hits(
+    @assert is_detector_hits_table(table) "Table does not have the correct format"
     
     # Collect all time-clustered results
     all_results = []
```
We should avoid declaring `Arrays` with unknown element-type, as this will default to `Any` and not be type-stable.
